5_Stable_implementation/flip.cpp

```diff
@@ -0,0 +1,332 @@
+
+#include <torch/torch.h>
+// #include <ATen/native/cpu/Loops.h>
+// #include <ATen/native/TensorAdvancedIndexing.h>
+
+
+struct AdvancedIndex {
+  AdvancedIndex(const torch::Tensor& src, torch::TensorList indices);
+
+  torch::Tensor src;
+  std::vector<torch::Tensor> indices;
+  torch::DimVector indexed_sizes;
+  torch::DimVector indexed_strides;
+  int64_t dims_before;
+  int64_t dims_after;
+};
+
+struct Indexer {
+  Indexer(int64_t num_indexers, char** indexers, const int64_t* indexer_strides,
+          torch::IntArrayRef original_sizes, torch::IntArrayRef original_strides)
+    : num_indexers(num_indexers)
+    , indexers(indexers)
+    , indexer_strides(indexer_strides)
+    , original_strides(original_strides.data())
+    , original_sizes(original_sizes.data()) {
+    AT_ASSERT(original_strides.size() == num_indexers);
+    AT_ASSERT(original_sizes.size() == num_indexers);
+  }
+
+  int64_t num_indexers;
+  char** indexers;
+  const int64_t* indexer_strides;
+  const int64_t* original_strides;
+  const int64_t* original_sizes;
+
+  int64_t get(int64_t idx) {
+    int64_t offset = 0;
+    for (int j = 0; j < num_indexers; j++) {
+      int64_t value = *(int64_t*)&indexers[j][idx * indexer_strides[j]];
+      offset += value * original_strides[j];
```

Comment:
You can optimize the code by moving the `original_strides` into the indices right away. This should bring some additional speedup as well